### PR TITLE
ENT-3474: Blackboard Client implementation

### PR DESCRIPTION
**Pull request by binodpant**
Does not include implementation of transmitter/exporters but does includes calls to content metadata exporter for now just to ensure things are wired up at least
Also skipping on some amount of testing since it will be duplicate effort post a planned refactor (to be done after blackboard integration basically works in stage)

In scope for this PR:
- [X] admin page for enterprise config
- [X] `/blackboard/oauth-complete` endpoint to get refresh token and save to database
- [X] client code to fetch token using the refresh token that will be used to make api calls
- [X] tests : light amount automated, mostly tested manually for now
 
Later down the line the PR to add exporter and transmitter will actually use the client access token

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [x] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [x] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/edx-enterprise), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

**Comment by binodpant** [comment]
FYI the code for blackboard and client are very very similar, with some minor differences, but very similar in terms of checking expiry etc. For now I am copying code quite a bit, but I am tempted at some point soon to refactor things out a bit so not going to invest in a lot of testing in blackboard client, because many of those tests will be reusable I think, I will likely do the detailed testing in a separate PR, once I do a bit of manual testing, I think PR will be good to go, for tomorrow. See you tomorrow!

File: integrated_channels/blackboard/client.py [hunk 1]
```diff
@@ -38,3 +43,104 @@ def create_course_completion(self, user_id, payload):  # pylint: disable=unused-
...
+            )
+            session = requests.Session()
+            session.headers['Authorization'] = 'Bearer {}'.format(oauth_access_token)
+            session.headers['Content-Type'] = 'application/json'
+            self.session = session
+            # expiry expected after `expires_in` seconds
+            if expires_in is not None:
+                self.expires_at = datetime.utcnow() + timedelta(seconds=expires_in)
+
+    def _get_oauth_access_token(self, client_id, client_secret):
```
Comment (binodpant) [inline integrated_channels/blackboard/client.py:69]: a bit embarrassed right now on the amount of copy pasting between this and canvas, but no worries because once this works well, I will be doing a refactor ticket, where I will also handle some tests for the client, so won't be going heavy on client test in this pr

**Comment by binodpant** [comment]
I will bump version before merge to avoid having to re-do it after rebase

File: integrated_channels/blackboard/client.py [hunk 2]
```diff
@@ -2,10 +2,20 @@
...
+
+import requests
+from six.moves.urllib.parse import urljoin  # pylint: disable=import-error
+
 from django.apps import apps
 
+from integrated_channels.exceptions import ClientError
 from integrated_channels.integrated_channel.client import IntegratedChannelApiClient
 
+# TODO: Refactor candidate (duplication with canvas client)
```
Comment (binodpant) [inline integrated_channels/blackboard/client.py:17]: these annoying comments to be removed when doing the refactor ticket

Comment (binodpant) [reply integrated_channels/blackboard/client.py:17]: annoying by design

File: integrated_channels/blackboard/admin/__init__.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+@admin.register(BlackboardEnterpriseCustomerConfiguration)
+class BlackboardEnterpriseCustomerConfigurationAdmin(admin.ModelAdmin):
+    """
+    Django admin model for BlackEnterpriseCustomerConfiguration.
+    """
+    list_display = (
+        "enterprise_customer_name",
+        "client_id",
+        "client_secret",
+        "blackboard_base_url",
```
Comment (taliaedX) [inline integrated_channels/blackboard/admin/__init__.py:19]: I haven't touched the admin pages here much, but why is enterprise_customer_name listed in `list_display` and readonly_fields, but `refresh_token` is only in read only?

File: integrated_channels/blackboard/client.py [hunk 4]
```diff
@@ -38,3 +50,97 @@ def create_course_completion(self, user_id, payload):  # pylint: disable=unused-
...
+                'Authorization': self._create_auth_header(),
+                'Content-Type': 'application/x-www-form-urlencoded'
+            }
+        )
+        if auth_response.status_code >= 400:
+            raise ClientError(auth_response.text, auth_response.status_code)
+        try:
+            data = auth_response.json()
+            # do not forget to save the new refresh token otherwise subsequent requests will fail
+            self.enterprise_configuration.refresh_token = data["refresh_token"]
```
Comment (taliaedX) [inline integrated_channels/blackboard/client.py:121]: In order to make requests you need a token, in order to get a token, you need the previous token, soooo what happens on the very first request? Given `refresh_token` is a read only field how do we get the initial token we need to get more tokens?

File: integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('blackboard', '0001_initial'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='blackboardenterprisecustomerconfiguration',
+            name='client_id',
```
Comment (taliaedX) [inline integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py:15]: where are we storing the token? Just in the session? Ie how would it be accessible to Django Admin if it's not in the db?

**Review by taliaedX — COMMENTED** [review]
I think I'm a little confused as to the authentication flow for Blackboard, but other than that this seems alright. Obviously needs a bit more fleshing out, but base level is fine. I'm just unsure how authentication even functions the way the code is written...

File: integrated_channels/blackboard/models.py [hunk 6]
```diff
@@ -25,20 +25,20 @@ class BlackboardEnterpriseCustomerConfiguration(EnterpriseCustomerPluginConfigur
...
-            "calls on behalf of the customer."
+            "calls on behalf of the customer. Called Application Key in Blackboard"
         )
     )
 
     client_secret = models.CharField(
         max_length=255,
         null=True,
-        verbose_name="API Client Secret",
+        verbose_name="API Client Secret or Application Secret",
```
Comment (alex-sheehan-edx) [inline integrated_channels/blackboard/models.py:38]: similar situation as to above- seems like this would be either the client secret or the application secret.

File: integrated_channels/blackboard/admin/__init__.py (hunk 3 above)
Comment (binodpant) [reply integrated_channels/blackboard/admin/__init__.py:19]: hmm, only 98% on this, but I believe enterprise_customer_name also needs to be shown in the changelist page which shows all the configs whereas the refresh token only displays once you select a config?

File: integrated_channels/blackboard/client.py (hunk 4 above)
Comment (binodpant) [reply integrated_channels/blackboard/client.py:121]: yes indeed, I think similar to canvas, we expect that there is a token the very first time we use this client, populated by following the manual workflow of using redirect_uri. Once that is in place, each request gets back a new refresh_token which I then store in the db for use next time, and so forth. but yeah first time won't work unless we have authorized and gotten the first refresh_token stored.

File: integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py (hunk 5 above)
Comment (binodpant) [reply integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py:15]: there is a `refresh_token` field in the db https://github.com/edx/edx-enterprise/pull/980/files#diff-444de76cae0d55485100349a6900b8e5R52 but we only store the refresh_token in the db

the access token is stored in the sessoin under 

`session.headers['Authorization'] = 'Bearer {}'.format(oauth_access_token)`

and used until it expires after which we refetch using that last fetched refresh_token stored in the db

File: integrated_channels/blackboard/client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/blackboard/client.py:121]: yup, that workflow of visit this url, authorize, will do that.

File: integrated_channels/blackboard/models.py [hunk 7]
```diff
@@ -25,20 +25,20 @@ class BlackboardEnterpriseCustomerConfiguration(EnterpriseCustomerPluginConfigur
     client_id = models.CharField(
         max_length=255,
         null=True,
-        verbose_name="API Client ID",
+        verbose_name="API Client ID or Blackboard Application Key",
```
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/blackboard/models.py:28]: yup I guess it could confuse a bit, would it be okay if I cleaned this up in the next coming pr? only because changing this would need to regen migration (instead of creating a new one I would need to clean up and regen I think?)

File: integrated_channels/blackboard/models.py (hunk 6 above)
Comment (binodpant) [reply integrated_channels/blackboard/models.py:38]: yup I guess it could confuse a bit, would it be okay if I cleaned this up in the next coming pr? only because changing this would need to regen migration (instead of creating a new one I would need to clean up and regen I think?)

File: integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py (hunk 5 above)
Comment (taliaedX) [reply integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py:15]: Hmm that diff doesn't seem to really work or at least in the files for this pr I do not see refresh token on the model nor the migration. Was it already defined in your previous pr? If so, you can ignore me here. :)

File: integrated_channels/blackboard/views.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            refresh_token = data['refresh_token']
+        except (KeyError, ValueError):
+            raise requests.RequestException(response=auth_response)
+
+        enterprise_config.refresh_token = refresh_token
+        enterprise_config.save()
+
+        return Response()
+
+    def _create_auth_header(self, enterprise_config):
```
Comment (alex-sheehan-edx) [inline integrated_channels/blackboard/views.py:120]: Seems like a very similar method to the one you have in the client, without the configuration check. Do you think it's worth making a common "create auth header" util? Not sure best practice here since it's such a minor function and only located in the two places (though perhaps larger use cases for integrated channels?)

File: tests/test_integrated_channels/test_blackboard/test_views.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+BAD_ENTERPRISE_ID = str(uuid4())
+SINGLE_CONFIG = {
+    'client_id': 'id',
+    'client_secret': 'secret',
+    'base_url': 'http://betatest.blackboard.com',
+}
+
+
+@pytest.mark.django_db
+class TestBlackboardAPIViews(APITestCase):
```
Comment (alex-sheehan-edx) [inline tests/test_integrated_channels/test_blackboard/test_views.py:32]: These are some good looking tests 😄

**Review by alex-sheehan-edx — COMMENTED** [review]
A few questions but overall looks pretty good to me.

File: integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py (hunk 5 above)
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py:15]: Just checked, looks like the refresh token field's already in master.

Comment (taliaedX) [reply integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py:15]: Thank you for doing what I was too lazy to do

File: integrated_channels/blackboard/views.py (hunk 8 above)
Comment (binodpant) [reply integrated_channels/blackboard/views.py:120]: yeah I had thought of doing that, left it in place right now only due to it's smallness

File: integrated_channels/blackboard/views.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        **Response Values**
+
+            HTTP 200 "OK" if successful
+
+            HTTP 400 if code/state is not provided
+
+            HTTP 404 if state is not valid or contained in the set of registered enterprises
+
+    """
+    def get(self, request, *args, **kwargs):  # pylint: disable=arguments-differ
```
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/blackboard/views.py:55]: yup, if we do ever hit it again it should be okay I thikn it just replaces with a new refresh token, I can test it though but yes that is the intent

File: integrated_channels/blackboard/client.py [hunk 11]
```diff
@@ -38,3 +50,97 @@ def create_course_completion(self, user_id, payload):  # pylint: disable=unused-
 
     def delete_course_completion(self, user_id, payload):  # pylint: disable=unused-argument
         """TODO"""
+
+    def _create_session(self):
```
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/blackboard/client.py:54]: hah yeah it's not used yet, since I did not have the transmitters etc to call. I was going to have the transmitter call the update* etc methods in my next PR where you will see this light up! I know it's odd, since this could be written now instead of new pr, but I thought it would be easier to test that change set together in transmitter pr

File: integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py (hunk 5 above)
Comment (binodpant) [reply integrated_channels/blackboard/migrations/0002_auto_20200930_1723.py:15]: yeah sorry should have explained that

File: integrated_channels/blackboard/views.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        app_config = apps.get_app_config('blackboard')
+        oauth_token_path = app_config.oauth_token_auth_path
+
+        # Check if encountered an error when generating the oauth code.
+        request_error = request.GET.get('error')
```
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/blackboard/views.py:60]: yeah I like it mainly for debugging times to keep that assignment line but not a strong feeling either way ...

File: tests/test_integrated_channels/test_blackboard/test_views.py (hunk 9 above)
Comment (binodpant) [reply tests/test_integrated_channels/test_blackboard/test_views.py:32]: good job Alex!

File: integrated_channels/blackboard/client.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/blackboard/client.py:54]: and yes no one nothing else is intended to call it other than that workflow

File: integrated_channels/blackboard/views.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/blackboard/views.py:55]: and yes no one nothing else is intended to call it other than that workflow

**Comment by binodpant** [comment]
> I think I'm a little confused as to the authentication flow for Blackboard, but other than that this seems alright. Obviously needs a bit more fleshing out, but base level is fine. I'm just unsure how authentication even functions the way the code is written...

Thanks Talia, indeed there are a couple confounding factors, I will be documenting some stuff in the confluence doc too which may be of help, but also some tests in upcoming PRs should help clarify this too I hope

File: integrated_channels/blackboard/views.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/views.py:60]: I think your impl is perfectly fine

File: integrated_channels/blackboard/client.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:54]: sounds good!

**Review by alex-sheehan-edx — APPROVED** [review]
LGTM! Thanks for addressing my comments/questions 🙂

**Review by taliaedX — APPROVED** [review]
LGTM (also just safe to merge regardless)
